hazelcast/src/main/java/com/hazelcast/jet/impl/JobCoordinationService.java

```diff
@@ -883,14 +942,17 @@ private <T> CompletableFuture<T> callWithJob(
         return submitToCoordinatorThread(() -> {
             // when job is finalized, actions happen in this order:
             // - JobResult and JobMetrics are created
-            // - JobRecord and JobExecutionRecord are deleted
+            // - JobRecord and JobExecutionRecord are deleted (asynchronously and in parallel)
             // - masterContext is removed from the map
             // We check them in reverse order so that no race is possible.
             //
-            // We check the JobResult after MasterContext for optimization because in most cases
-            // there will either be MasterContext or JobResult. Neither of them is present only after
-            // master failed and the new master didn't yet scan jobs. We check the JobResult
-            // again at the end for correctness.
+            // We check the JobResult after MasterContext for correctness and
```

Comment:
@viliam-durina  this order of check is necessary for correctness, not only for optimization

Reply:
`JobResult` after MC is optimization (getting MC is cheaper because it is ordinary local map) but I think that `JobResult` before `JobRecord` and `JobExecutionRecord` is necessary for correctness or at least less eventually-consistent behavior. `JobResult` and `JobRecord` can both exist when the MC is already deleted. It such case we should use `JobResult` because `JobRecord` can be outdated and does not contain all information (eg. user-cancellation flag).

Reply:
updated